src/HexagonOptimize.cpp

```diff
@@ -1147,6 +1219,219 @@ class OptimizePatterns : public IRMutator {
+                IRMatcher::Wild<0> x;
+                IRMatcher::Wild<1> y;
+                auto rewrite = IRMatcher::rewriter(IRMatcher::mul(op->a, op->b), op->type);
+                if (rewrite(h_add(x, lanes) * broadcast(y, lanes), h_add(x * broadcast(y, arg_lanes), lanes)) ||
+                    rewrite(h_add(y, lanes) * broadcast(x, lanes), h_add(broadcast(x, arg_lanes) * y, lanes))) {
+                    return mutate(rewrite.result);
+                }
+            }
+        }
+        return IRMutator::visit(op);
+    }
+
+    typedef struct {
+        enum Flags {
+            SlidingWindow = 1,
+            ScalarB = 1 << 1
+        };
+        int rfac;
+        Type r_ty;
+        Type a_ty;
+        Type b_ty;
+        int flags;
+    } Signature;
+
+    Expr visit(const VectorReduce *op) override {
+        if (op->type.bits() == 8 || !op->type.is_vector() || op->type.is_float() || op->op != VectorReduce::Add) {
+            return IRMutator::visit(op);
+        }
+        int in_lanes = op->value.type().lanes();
+        int out_lanes = op->type.lanes();
+        int rfac = in_lanes / out_lanes;
+
+        // Map of instruction signatures
+        static const vector<Signature> sigs = {
+            // --------- vrmpy ---------
+            // Sliding window
+            {4, UInt(32, 0), UInt(8, 0), UInt(8, 4), Signature::SlidingWindow | Signature::ScalarB},
+            {4, Int(32, 0), UInt(8, 0), Int(8, 4), Signature::SlidingWindow | Signature::ScalarB},
+            // Vector * Scalar
+            {4, UInt(32, 0), UInt(8, 0), UInt(8, 4), Signature::ScalarB},
+            {4, Int(32, 0), Int(8, 0), UInt(8, 4), Signature::ScalarB},
+            {4, Int(32, 0), UInt(8, 0), Int(8, 4), Signature::ScalarB},
+            // Vector * Vector
+            {4, UInt(32, 0), UInt(8, 0), UInt(8, 32)},
+            {4, Int(32, 0), UInt(8, 0), Int(8, 32)},
+            {4, Int(32, 0), Int(8, 0), Int(8, 32)},
+
+            // --------- vtmpy ---------
+            // Sliding window
+            {3, Int(16, 0), Int(8, 0), Int(8, 3), Signature::SlidingWindow | Signature::ScalarB},
+            {3, Int(16, 0), UInt(8, 0), Int(8, 3), Signature::SlidingWindow | Signature::ScalarB},
+            {3, Int(32, 0), Int(16, 0), Int(8, 3), Signature::SlidingWindow | Signature::ScalarB},
+
+            // --------- vdmpy ---------
+            // Sliding window
+            {2, Int(16, 0), UInt(8, 0), Int(8, 4), Signature::SlidingWindow | Signature::ScalarB},
+            {2, Int(32, 0), Int(16, 0), Int(8, 4), Signature::SlidingWindow | Signature::ScalarB},
+            // Vector * Scalar
+            {2, Int(32, 0), Int(16, 0), UInt(16, 2), Signature::ScalarB},  // Saturates
+            {2, Int(32, 0), Int(16, 0), Int(16, 2), Signature::ScalarB},   // Saturates
+            {2, Int(16, 0), UInt(8, 0), Int(8, 2), Signature::ScalarB},    // b_ty can be Int(8, 4).
+            // TODO: This would not match untill b_ty is changed to Int(8, 4)
```

Comment:
It seems like this should be Int(8, 2), but the call to the intrinsic should broadcast it 2x? i.e. does the TODO belong somewhere else?

Reply:
If we use Int(8, 4) instead of broadcasting to 2x, we can match instructions for patterns like:
a*v0 + b*v1,  c*v2 + d*v3, a*v4 + b*v5
as well